content/learn/05.communication/07.modbus/modbus.md

```diff
@@ -0,0 +1,119 @@
+
+Modbus requests follow a simple structure:
+![](assets/requestFrame.png)
+
+### Device Address
+Every peripheral device has its own address which it responds to when addressed by the controller, while all other devices ignore the message if the address doesn't match their own. Device addresses are assigned in the range of 1 - 247, but without additional hardware the maximum number of devices should be limited to 32 as data loss is too significant when attaching more devices. 
+
+### Function Code
+The Function code tells the peripheral device if it should read or write data from the internal memory registers. Many of the data types are named from their use in driving relays, for example, a single-bit physical output is called a coil, and a single-bit physical input is called a discrete input or a contact. The following functions are supported by the Modbus poll:
+
+- 01 READ COIL
+- 02 READ DISCRETE INPUT
+- 03 READ HOLDING REGISTERS
+- 04 READ INPUT REGISTERS
+- 05 WRITE SINGLE COIL
+- 06 WRITE SINGLE REGISTER
+- 15 WRITE MULTIPLE COILS
+- 16 WRITE MULTIPLE REGISTERS
+
+### Data
+The data field contains the requested or send data. In the request form used by the Arduino Modbus library, the data field requires you to enter the starting registers and register count.
+
+### CRC Error Check
+CRC stands for Cyclic Redundancy check and is an error detecting code commonly used in digital networks to detect accidental changes in digital data. CRC’s have their name because the check (data verification) value is a redundancy (it expands the message without adding information) and the algorithm is based on [cyclic codes](https://en.wikipedia.org/wiki/Cyclic_code#:~:text=In%20coding%20theory%2C%20a%20cyclic,efficient%20error%20detection%20and%20correction.). They are simple to implement in binary hardware, easy to analyze mathematically, and particularly good at detecting common errors caused by noise in transmission channels. 
+
+Two bytes are added to the end of every Modbus message for error detection and every byte in the message is used to calculate the CRC. The receiving device then also calculates the CRC and compares it to the CRC from the sending device. If even one bit in the message is received incorrectly, the CRC’s will be different resulting in an error.
+
+### Byte and word ordering
+The Modbus specification doesn't define exactly how the data is stored in the registers. Therefore, some manufacturers implemented Modbus in their equipment to store and transmit the higher byte first followed by the lower byte.  Alternatively, others store and transmit the lower byte first. Similarly, when registers are combined to represent 32-bit data types, some devices store the higher 16 bits (high word) in the first register and the remaining low word in the second while others do the opposite. It doesn't matter which order the bytes or words are sent in, as long as the receiving device knows which way to expect it.
+
```

Comment:
For this Byte and word Ordering section, as it is using the words from a source, it would need proper citation due to its explicit reference.